### ISSUE TITLE
Error when building item features using dataset construction

### ISSUE THREAD

**Issue by cyteen** [lyst/lightfm#378]
I have mutiple features for item that I'm trying to include in the model.
I followed each steps as shown in the examples and documentation.
However, when trying to build item features using the dataset.build_item_features, I get error 
'Feature automobile not in feature mapping. Call fit first.'

Subset of my dataframe ( name: ad):

 ad_id | ad_category |  ad_size  | ad_format
-------|--------------|-----------|----------------
1356  | automobile  | 336x280 | image
3455  | food             | 250x250 | text
2344  | food             | 200x200 | text

My codes are shown below. 

#### Create dataset  
    
    dataset = Dataset()
    dataset.fit(users= user['cuuid'].unique(), 
                     items= ad['ad_id'].unique(),
                     item_features= [tuple(i) for i in ad.drop(['ad_id'], axis = 1).values])
    #### I created an iterable of my item features in a tuple form. (Question 1)

This is a subset on how my item features in tuple form looks like:
[('automobile', '336 x 280', 'image'),
 ('food', '250 x 250', 'text'),
 ('food', '200 x 200', 'text')]

#### Building user-item matrix
    # Created (user id, item id, click)
    action= [tuple(i) for i in merged.drop(['click','view','ad_category', 'ad_size', 'ad_format'], axis = 
                  1).values]
     interactions, weights = dataset.build_interactions(action)

A subset on how my action tuple looks like. (userid, itemid, click or not)
[(1, 1356, 1),
 (2, 3455, 1),
 (3, 2344, 1)]

#### Building item feature matrix
    ad_subset = ad[["ad_category", 'ad_size','ad_format']] 
    ad_list = [list(x) for x in ad_subset.values]
    ad_tuple = tuple(zip(ad['ad_id'], ad_list))
    (Question 2)

A subset of the ad_tuple
((1356, ['automobile', '336 x 280', 'image']),
 (3455, ['food', '250 x 250', 'text']),
 (2344, ['food', '200 x 200', 'text']))

    item_features = dataset.build_item_features(ad_tuple)

#### Fit model
    model = LightFM(loss='warp')
    model.fit(interactions, 
          item_features= item_features,
          sample_weight= weights,
          epochs=1, 
          num_threads=2)

Question 1: Iterating my item features in tuple form. Is this the correct way? 
Question 2: Is the format of my ad_tuple correct?
Question 3: Since my features are categorical, should I one-hot encode it?

**Comment by maciejkula** [comment]
Thanks, @DoronGi!

**Comment by cyteen** [comment]
My mistake was inputing item features in Dataset in the form of tuples by tuples instead of all features in 1 long list. 
It works now. 
Thank you @DoronGi

**Comment by ctivanovich** [comment]
What a crazy input method.
